### PR TITLE
task/WP-156-Exceptions-Admin-Table-UTH-Requests

### PR DESCRIPTION
**Pull request by sophia-massie**
## Overview
Adding filters, action dropdown and edit modal for admin exception table

## Related

- [WP-156](https://jira.tacc.utexas.edu/browse/WP-156)

## Changes

- Added status and organization filters and url paths for query params
- Added action dropdown to view and edit exceptions
- To edit an exception, created: 

> 1. Edit exception modal template
> 2. update_exception query
> 3. Modified modal styles for extension
> 4. Success and error templates
> 5. post function in views.py
- Reordered view exception per requests
- Made sure when page is narrowed, the column names align with the new row names

## Testing

1. Go to [http://localhost:8000/submissions/other-exception/](http://localhost:8000/submissions/other-exception/) and submit an exception that will then be stored to the test db.
2. Go to [http://localhost:8000/administration/list-exceptions/?org=All](http://localhost:8000/administration/list-exceptions/)
3. The orgs and statuses are being pulled from the DB directly and not hardcoded in the view. Once the db is updated, it will put all status and orgs available.
- CAREFUL: Try it with the prod DB creds to be sure  the filters are populated correctly. Do not to edit any exception entries that are not tests.
4. Check the org and status filter independently and make sure the entries that display are correct through all pages
5. Try them together and make sure the entries that display are correct through all pages.
6. Try to edit an exception. All fields that you can edit should populate with what is in the DB for that exception_id. 
7. Not all fields are required in the edit modal. Make sure anything you did not change in the edit modal stayed the same after submission.
8. If the exception is a threshold request, the edit modal should allow you to edit the approved threshold % and show the requested threshold %.
9. Check the existing exception information at the bottom of the edit modal. Any field that is not editable at the top should be displayed at the bottom in the same order as the view modal. 
10. Check that the view modal does not have duplicate fields listed

## UI
Exception Admin Table
Screenshot 2023-06-26 at 5 15 50 PM

View Exception Modal
Screenshot 2023-06-26 at 5 16 03 PM

Status filter  Org filter will only show Tacc Test org on test db
Screenshot 2023-06-26 at 5 16 22 PM

Edit exception modal pt 1
Screenshot 2023-06-26 at 5 16 53 PM

If editing a threshold exception, should show threshold field
Screenshot 2023-06-26 at 5 17 38 PM

Edit exception modal pt 2

Screenshot 2023-06-26 at 5 17 51 PM

### CODE REVIEW

File: apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html [hunk 1]
```diff
@@ -2,22 +2,49 @@
 {% load static %}
 
 {% block content %}
-
 <link rel="stylesheet" href="{% static 'apcd-cms/css/table.css' %}">
 <link rel="stylesheet" href="{% static 'apcd-cms/css/modal.css' %}">
-<link rel="stylesheet" href="{% static 'submissions/css/table.css' %}">
+<!-- Must come before submissions modal stylesheet so key names in modal are not cut off -->
+<link rel="stylesheet" href="{% static 'admin_exception/css/table.css' %}">
```
Comment (edmondsgarrett) [inline apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html:8]: Filter drop downs on this page are compressed right around the selected option; Wes cleaned up the filter styles really nicely, I think you can source these styles instead like is done in the admin registrations table:

`<link rel="stylesheet" href="{% static 'apcd-cms/css/table.css' %}">`

File: apcd-cms/src/apps/admin_exception/templates/view_admin_exception_modal.html [hunk 2]
```diff
@@ -15,39 +15,43 @@ <h4 class="modal-title">Exception Details</h4>
...
+                      <dt class="c-data-list__key">Created</dt>
+                      <dd class="c-data-list__value">{{r.created_at}}</dd>
+                      <dt class="c-data-list__key">Organization</dt>
+                      <dd class="c-data-list__value">{{r.org_name}}</dd>
+                      <dt class="c-data-list__key">Requestor</dt>
+                      <dd class="c-data-list__value">{{r.requestor_name}}</dd>
+                      <dt class="c-data-list__key">Requestor Email</dt>
+                      <dd class="c-data-list__value">{{r.requestor_email}}</dd>
+                      <dt class="c-data-list__key">Exception Type</dt>
+                      <dd class="c-data-list__value">{{r.exception_type}}</dd>
```
Comment (edmondsgarrett) [inline apcd-cms/src/apps/admin_exception/templates/view_admin_exception_modal.html:27]:
```suggestion
                      <dd class="c-data-list__value">{{r.request_type}}</dd>  
```
Aligning this var with that in the view.

File: apcd-cms/src/apps/admin_exception/templates/edit_exception_modal.html [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                    <dt class="c-data-list__key">Created</dt>
+                    <dd class="c-data-list__value">{{r.created_at}}</dd>
+                    <dt class="c-data-list__key">Organization</dt>
+                    <dd class="c-data-list__value">{{r.org_name}}</dd>
+                    <dt class="c-data-list__key">Requestor</dt>
+                    <dd class="c-data-list__value">{{r.requestor_name}}</dd>
+                    <dt class="c-data-list__key">Requestor Email</dt>
+                    <dd class="c-data-list__value">{{r.requestor_email}}</dd>
+                    <dt class="c-data-list__key">Exception Type</dt>
+                    <dd class="c-data-list__value">{{r.exception_type}}</dd>
```
Comment (edmondsgarrett) [inline apcd-cms/src/apps/admin_exception/templates/edit_exception_modal.html:102]:
```suggestion
                    <dd class="c-data-list__value">{{r.request_type}}</dd>  
```
Aligning this with var defined in view.

File: apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html [hunk 4]
```diff
@@ -2,22 +2,49 @@
...
+      </div>
+      </div>
+      {% if selected_status or selected_org %}
+      <div class="filter-container">
+        <div class="filter-content">
+          <button onclick="clearSelections()">Clear Options</button>
+        </div>
+      </div>
+      {% endif %}
+  <hr/>
```
Comment (edmondsgarrett) [inline apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html:46]: I suggest moving this `<hr>` element to before the filter container to align with design on other listing pages.

**Review by edmondsgarrett — CHANGES_REQUESTED** [review]
Modals look sleek and work well, good job! I have some comments here and below:

I think having the filters stacked on top of each other above the table here may be making these filters take up too much room on the page - I also get the appeal of doing this so filters do not stretch all the way across the width of the table. Not sure which way is better, maybe reduce vertical space between filters?

File: apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html (hunk 1 above)
Comment (sophia-massie) [reply apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html]: Updated! The only issue is that the submissions table style sheet is the only thing that keeps the modal from cutting off the labels for each row. I moved both the modal and submission modal style sheet to the modal only for now until the global modal css addresses this.

File: apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html (hunk 4 above)
Comment (sophia-massie) [reply apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html]: Good idea! Made the filters more in line with Wes' styles and moved the hr to match other pages

File: apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html (hunk 1 above)
Comment (sophia-massie) [reply apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html]: Let me know if you'd like any other changes on the modal style sheets

Comment (sophia-massie) [reply apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html]: I'll actually remove any applicable styles that help the modal render without cutting off text in the data list from submissions to the admin_exception stylesheet. Working on this now

Comment (sophia-massie) [reply apcd-cms/src/apps/admin_exception/templates/list_admin_exception.html:6]: This has been updated. Thanks for the suggestion!

**Review by sophia-massie — COMMENTED** [review]
Should be ready for another look. Thanks for your help, Garrett!

**Review by fnets — APPROVED** [review]
LGTM!

**Pull request merged**